### PR TITLE
Feature: Convert Grafana dashboards to KubeSphere monitoring templates

### LINKED ISSUE #7: Feature: Convert Grafana dashboards to KubeSphere monitoring templates
Grafana is a successful and widely used observability tool.  In future releases, we should support importing dashboards from Grafana dashboard templates. It can be implemented with a binary handing this conversion.

### PR DESCRIPTION
**Pull request by zhu733756**
fix #7, features are described as following:
-  a new version v1alpha2 for crd has provided, which supports more fields, refers to README.md.
- the new version and the older version are compatible with each other .
- add a converter tool, aims to help for importing grafana dashboard templates, and convet to kubesphere dashboard manifests.

### CODE REVIEW

File: api/v1alpha1/panels/dashboard_penels_types.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
+// +kubebuilder:object:generate=true
```
Comment (benjaminhuo) [inline api/v1alpha1/panels/dashboard_penels_types.go]: dashboard_penels_types.go  should be dashboard_panels_types.go  ?

File: api/v1alpha1/panels/bargauge_types.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// +kubebuilder:object:generate=true
+package panels
+
+// refers to https://pkg.go.dev/github.com/grafana-tools/sdk#BarGaugePanel
+type BarGaugePanel struct {
+	Options BarGaugeOptions `json:"options,omitempty,omitempty" yaml:"options,omitempty,omitempty"`
```
Comment (benjaminhuo) [inline api/v1alpha1/panels/bargauge_types.go]: if the member of a struct is of type struct, then this member should be a pointer to a struct type.
pls check all relevant code

File: config/prometheus/es-daemonset.yaml [hunk 3]
```diff
@@ -0,0 +1,63 @@
+piVersion: extensions/v1beta1
+kind: DaemonSet
```
Comment (benjaminhuo) [inline config/prometheus/es-daemonset.yaml]: These files are not necessary:

config/prometheus/grafana-svc.yaml
config/prometheus/grafana-deploy.yaml
config/prometheus/es_servicemonitor.yaml
config/prometheus/es-daemonset.yaml

File: tools/converter/json.go [hunk 4]
```diff
@@ -0,0 +1,785 @@
...
+		kind = "ClusterDashboard"
+	}
+
+	if ns == "" {
+		ns = "default"
+	}
+	manifest := k8sDashboard{
+		APIVersion: apiVersion,
+		Kind:       kind,
+		Metadata:   map[string]string{"name": name, "namespace": ns},
```
Comment (benjaminhuo) [inline tools/converter/json.go]: ClusterDashboard shouldn't have a namespace

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -1,183 +1,214 @@
...
+### Panels
+
+#### Chart
+
+We currently support four types of panels
+
+- Row
+- Singlestat
+- Graph
+
```
Comment (junotx) [inline README.md]: Only three of the four supported panel types are listed

File: Makefile [hunk 6]
```diff
@@ -1,80 +1,89 @@
...
+# find or download controller-gen
+# download controller-gen if necessary
+controller-gen:
+ifeq (, $(shell which controller-gen))
+	@{ \
+	set -e ;\
+	CONTROLLER_GEN_TMP_DIR=$$(mktemp -d) ;\
+	cd $$CONTROLLER_GEN_TMP_DIR ;\
+	go mod init tmp ;\
+	go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.2.5 ;\
```
Comment (junotx) [inline Makefile]: This version of the controller-gen generated CRD version for `apiextensions.k8s.io/v1beta1`, but the latest code has upgraded it to v0.4.1 which generates CRDs for the `apiextensions.k8s.io/v1` version required by KubeSphere3.1+

File: api/v1alpha1/panels/dashboard_penels_types.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+// DashboardPanel referers to https://pkg.go.dev/github.com/K-Phoen/grabana/decoder#DashboardPanel
+// type DashboardPanel struct {
+// 	Name   string  `json:"name,omitempty" yaml:"name,omitempty"`
+// 	Repeat string  `json:"repeat,omitempty" yaml:"repeat,omitempty"`
+// 	Panels []Panel `json:"panels,omitempty" yaml:"panels,omitempty"`
+// }
+
+// Supported panel type
+type Panel struct {
```
Comment (junotx) [inline api/v1alpha1/panels/dashboard_penels_types.go]: It looks like `Panel` already supports configuring multiple types of panels, but why are there still other definitions like `Graph`, `Table`, `Text` and so on.

**Review by junotx — CHANGES_REQUESTED** [review]

**Comment by junotx** [comment]
The conflicts can also be resolved, and compatibility with previous original configuration is recommended, which can be verified by unit testing @zhu733756

File: api/v1alpha1/panels/dashboard_penels_types.go (hunk 7 above)
Comment (zhu733756) [reply api/v1alpha1/panels/dashboard_penels_types.go]: Have got two question as followings:
- Currently，we only support panel types like `graph/row/singlestat`,  the other types no need to support?
- The detailed definition of the panel types no need to convert to the dashboard? The style of the console was a  fixed struct?

File: Makefile (hunk 6 above)
Comment (zhu733756) [reply Makefile]: Would be fixed

**Comment by zhu733756** [comment]
> The conflicts can also be resolved, and compatibility with previous original configuration is recommended, which can be verified by unit testing @zhu733756

Would be fixed. Pls answer my two questions above and  I would readjust it according to your suggestion.

File: api/v1alpha1/panels/dashboard_penels_types.go (hunk 7 above)
Comment (junotx) [reply api/v1alpha1/panels/dashboard_penels_types.go]: For the first, other types of support can be added. And the second, i don't quite understand what you mean

File: api/v1alpha1/dashboard_types.go [hunk 8]
```diff
@@ -64,6 +64,44 @@ type Panel struct {
...
+// 	switch {
+// 	case p.Row != nil:
+// 		return json.Marshal(p.Row)
+// 	case p.Graph != nil:
+// 		return json.Marshal(p.Graph)
+// 	case p.SingleStat != nil:
+// 		return json.Marshal(p.SingleStat)
+// 	}
+// 	return json.Marshal(p)
+// }
```
Comment (junotx) [inline api/v1alpha1/dashboard_types.go]: These can be deleted directly instead of comments

File: api/v1alpha2/dashboard_conversion.go [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+	dst.Spec.Panels = pls
+
+	for _, temp := range src.Spec.Templatings {
+		dst.Spec.Templatings = append(dst.Spec.Templatings, v1alpha1.Templating{
+			Name:  temp.Name,
+			Query: temp.Request,
+		})
+	}
+
```
Comment (junotx) [inline api/v1alpha2/dashboard_conversion.go]: Why is the `request` attribute used, and i counld not find it in the grafana template struct

File: api/v1alpha2/panels/dashboard_penels_types.go [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+			// var row RowPanel
+			// if err = json.Unmarshal(b, &row); err == nil {
+			// 	p.RowPanel = &row
+			// }
+			// default:
+			// 	var custom = make(CustomPanel)
+			// 	if err = json.Unmarshal(b, &custom); err == nil {
+			// 		p.CustomPanel = &custom
+			// 	}
+		}
```
Comment (junotx) [inline api/v1alpha2/panels/dashboard_penels_types.go]: `MarshalJSON` has handling for `row` case, but not here

**Review by junotx — CHANGES_REQUESTED** [review]

File: api/v1alpha1/dashboard_types.go (hunk 8 above)
Comment (zhu733756) [reply api/v1alpha1/dashboard_types.go]: would be fixed

File: api/v1alpha2/panels/dashboard_penels_types.go (hunk 10 above)
Comment (zhu733756) [reply api/v1alpha2/panels/dashboard_penels_types.go]: Row struct is empty, these comments would be used for feature if we want to add some fields.

File: api/v1alpha2/dashboard_conversion.go (hunk 9 above)
Comment (zhu733756) [reply api/v1alpha2/dashboard_conversion.go]: it refers to [the dark definition]( https://pkg.go.dev/github.com/K-Phoen/grabana/decoder#DashboardVariable),  would be fixed.

**Review by junotx — APPROVED** [review]
/lgtm

**Pull request merged**